onprc_ehr/src/org/labkey/onprc_ehr/issues/RestrictedIssueProviderImpl.java

```diff
@@ -0,0 +1,136 @@
+    private static final String PROP_RESTRICTED_ISSUE_LIST = "issueRestrictedIssueList";
+    private static final String PROP_RESTRICTED_ISSUE_LIST_GROUP = "issueRestrictedIssueListGroup";
+
+    @Override
+    public void setRestrictedIssueTracker(Container c, String issueDefName, Boolean isRestricted)
+    {
+        setPropertyValue(c, issueDefName, PROP_RESTRICTED_ISSUE_LIST, String.valueOf(isRestricted));
+    }
+
+    @Override
+    public boolean isRestrictedIssueTracker(Container c, String issueDefName)
+    {
+        String value = getPropertyValue(c, issueDefName, PROP_RESTRICTED_ISSUE_LIST);
+        if (value != null)
+        {
+            return Boolean.parseBoolean(value);
+        }
+        return false;
+    }
+
+    @Override
+    public void setRestrictedIssueListGroup(Container c, String issueDefName, @Nullable Group group)
+    {
+        setPropertyValue(c, issueDefName, PROP_RESTRICTED_ISSUE_LIST_GROUP, null != group ? String.valueOf(group.getUserId()) : "0");
+    }
+
+    @Override
+    public @Nullable Group getRestrictedIssueListGroup(Container c, String issueDefName)
+    {
+        String groupId = getPropertyValue(c, issueDefName, PROP_RESTRICTED_ISSUE_LIST_GROUP);
+        if (null == groupId)
+            return null;
+
+        return SecurityManager.getGroup(Integer.valueOf(groupId));
+    }
+
+    private void setPropertyValue(Container c, String issueDefName, String key, String value)
+    {
+        PropertyManager.PropertyMap props = PropertyManager.getWritableProperties(c, getPropMapName(issueDefName), true);
+        props.put(key, value);
+        props.save();
+    }
+
+    private String getPropertyValue(Container c, String issueDefName, String key)
+    {
+        return PropertyManager.getProperties(c, getPropMapName(issueDefName)).get(key);
+    }
+
+    private String getPropMapName(String issueDefName)
+    {
+        if (issueDefName == null)
+            throw new IllegalArgumentException("Issue def name must be specified");
+
+        return CAT_RESTRICTED_ISSUE_DEF_PROPERTIES + issueDefName;
+    }
+
+    @Override
+    public boolean hasPermission(User user, @NotNull Issue issue, List<Issue> relatedIssues, List<ValidationError> errors)
+    {
+        Container issueContainer = ContainerManager.getForId(issue.getContainerId());
+        if (issueContainer != null && isRestrictedIssueTracker(issueContainer, issue.getIssueDefName()))
+        {
+            Group group = getRestrictedIssueListGroup(issueContainer, issue.getIssueDefName());
+            if (!checkAccess(user, issue, group))
+            {
+                errors.add(new SimpleValidationError("This issue is in a restricted issue list. You do not have access to this issue"));
+                return false;
+            }
+        }
+
+        // the user must also have access to all related issues
+        for (Issue related : relatedIssues)
```

Comment:
I saw in the platform PR that before calling this `hasPermission` method, it gathers up the `relatedIssues` and passes them through to here. Thoughts on having this method be the one that gets the related issues instead? Is there a scenario where this method is called with a different set of related issues then the ones from the IssueManager call?

Reply:
Yes that would have looked cleaner but there wasn't a great way to maintain a single choke point with `IssueManager.getIssue` in a way to do access checking in one place. `IssueManager.getIssue` calls the helper method `_getIssue` to get the fully materialized issue and I didn't want to expose that as a public method to allow callers to get an issue without checking or add a flag to suppress access checking.